### DISCUSSION TITLE
Is it possible to combine projections for constrained optimization?

### DISCUSSION THREAD

**Discussion by vivien000** [google/jaxopt#247] [Q&A] [answered]
Hi! I'd like to solve a constrained optimization problem like:
$$\min f(x)$$
with:
- $x \in \mathcal{R}^{n+}$
- $Ax = b$

I understand it's possible to enforce the first constraint with `jaxopt.projection.projection_non_negative` and the second one with `jaxopt.projection.projection_affine_set`. But is it possible to enforce both at the same time? If so, how?

Many thanks in advance and thanks for the great library!

**Comment by vivien000** [comment]
After having thought about it, if I get it right:
- for the example I gave above, we can just use `jaxopt.projection.projection_polyhedron`
- in general, we can enforce as many constraints as we like as long as we're able to write the corresponding projection such as in [projection.py](https://jaxopt.github.io/stable/_modules/jaxopt/_src/projection.html#projection_polyhedron)

**Comment by mblondel** [comment, marked as answer]
If `f` is a quadratic function, you can use one of our [quadratic programming](https://jaxopt.github.io/stable/quadratic_programming.html) solvers. If `f` is a more general function, projected gradient + projection_polyhedron would be indeed one way to do it, although the projection needed on each iteration could be expensive to compute. We don't support it in the library yet but ADMM comes to mind for solving this type of problem.
